src/xlat-i/xi-Entry.C

```diff
@@ -1870,21 +1914,51 @@ void Entry::genRegularCall(XStr& str, const XStr& preCall, bool redn_wrapper, bo
       if(isRdmaPost) {
         str << "#if CMK_ONESIDED_IMPL\n";
         // Allocate an array of rdma pointers
-        str << "  void *buffPtrs["<< numRdmaRecvParams <<"];\n";
-        str << "  int buffSizes["<< numRdmaRecvParams <<"];\n";
+        if (param->hasDevice()) {
+          str << "  void *buffPtrs["<< numRdmaRecvParams + numRdmaDeviceParams <<"];\n";
+          str << "  int buffSizes["<< numRdmaRecvParams + numRdmaDeviceParams <<"];\n";
+          str << "#else\n";
+          str << "  void *buffPtrs["<< numRdmaDeviceParams <<"];\n";
```

Comment:
What's the plan for supporting ZC Post Device API on non-RDMA layers? (like netlrts). Is that why this code was added? (the two lines after `#else`). I am guessing that you can always support Cuda IPC transfers as long as you have GPU, but support for GPU direct is limited to only a few networks?

Reply:
I think the device-side ZC API should be supported also for non-RDMA layers (i.e. `CMK_ONESIDED_IMPL` disabled), which is why there have been some changes to accomodate this in the code. Adding in GPUDirect RDMA for inter-node GPU messaging will make things more complicated, because the architectures that support it may not match the ones that support host-side RDMA. I'm leaving that for the future me to mull over, for now.